design/cryptography.md

```diff
@@ -0,0 +1,127 @@
+- The Public key will be published to the Management contract and used by all obscuro tooling ( like the wallet extension ) to encrypt transactions
+- The Private key will be determined by other enclaves by deriving the Master Seed + Genesis Rollup 
+  - The Master Seed will be shared using the attestation process
+  - The Genesis Rollup will be shared when reading the L1 blocks
+- HKDF (HMAC-based Key Derivation Function) is used to derive keys
+  - Given the high entropy of the Master Seed no need for PBKDF Family key stretching
+  - Derivations use public rollup metadata such as height ( or some other shared field ) in the Info component of HKDF and use a fixed size salt as per https://soatok.blog/2021/11/17/understanding-hkdf/
+
+
+## 3. Encrypting transactions such that they can be revealed
+- Transactions are encrypted by default using the smallest unit of time revelation period
+- Transactions using AccessList can specify the desired revelation period
+- Using Null Address combined with hexadecimal 1-5 level determines the revelation period
+- Obscuro interprets the revelation period and encrypts the transaction with the corresponding revelation period key
+
+The [specification](https://eips.ethereum.org/EIPS/eip-2930) of the `AccessList` Field is as follows:
+
+```
+For the transaction to be valid, accessList must be of type [[{20 bytes}, [{32 bytes}...]]...], 
+where ... means “zero or more of the thing to the left”.
+For example, the following is a valid access list (all hex strings would in reality be in byte representation):
+
+[
+    [
+        "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",
+        [
+            "0x0000000000000000000000000000000000000000000000000000000000000003",
+            "0x0000000000000000000000000000000000000000000000000000000000000007"
+        ]
+    ],
+    [
+        "0xbb9bc244d798123fde783fcc1c72d3bb8c189413",
+        []
+    ]
+]
+```
+
+Obscuro uses the `AccessList` object to determine the revelation period.
+A special address and special storage keys combination is checked for internally.
+
+To simplify the protocol the Null address (0x0000000000000000000000000000000000000000) will be used and the hexadecimal representation of 1 to 5 is used to specify the revelation period.
+
+An example would be:
+```
+[
+    [
+        "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",                             <- Canonical AccessList usage
+        [
+            "0x0000000000000000000000000000000000000000000000000000000000000003",
+            "0x0000000000000000000000000000000000000000000000000000000000000007"
+        ]
+    ],
+    [
+        "0x0000000000000000000000000000000000000000",                             <- Obscuro Revelation Period
+         [
+            "0x0000000000000000000000000000000000000000000000000000000000000003", <- 1 day revelation Period
+            "0x0000000000000000000000000000000000000000000000000000000000000007"  <- ignored
+        ]
+    ]
+]
+```
```

Comment:
There's a few assumptions that aren't right. Hopefully, with the document review this becomes clearer.

- The privacy/encryption is at the transaction level not the smart contract level.
   The user issues a transaction and that transaction can call 1000 contracts but it's all encrypted for the defined transaction revelation period.
   
 - To avoid disruption to current web3 frameworks, this will reuse the Access List field with specific data.
     Obscuro parses the Access List fields and encrypts with the corresponding Revelation key.

Hope that cleares up a bit things. To answer the questions:

> how will that be done in practice,since the standard ethereum tooling has no clue about the revelation? At what level will this be added in practice?

Short example answer: User goes to Obscuro-Uniswap site, User requests a swap AND selects a Revelation Period of 1 year, Obscuro-Uniswap ethers framework creates the transaction AND adds `[0x0..., [0x0..04]]` in the Accesslist Field, bla bla, Obscuro enclave unpacks the transaction, checks the AccessList field, sees `Null Address + Revelation Period 4` and encrypts the transaction with the 1 year encryption revelation key.

>What if a tool builds a "create" transaction that uses the AccessList? How will that play with the revelation using it?

Good q. The AccessList is a list of addresses, we're just appending one entry with the null address.

So if the a transaction with really specific AccessList items is created AND it requires a 1 month revelation period, it would just append that info to the list.